File: mtree_replication.py
#!/usr/bin/env python

import pexpect
import sys
import re
import os

'''This script creates mtrees replication between two Data Domain storages'''

#----EDIT THIS-----------VVVV
#source:

mtree_name='JJJJ'             #Mtree name - a non-existing one
source = 'x.x.x.x'     #external dd management address for source
username = 'xxxx'         #login name in DD for IP: 10.246.168.131
pswd = 'xxxx'             #password for the above login name

#destination:
destination ='x.x.x.x' #external dd management address for destination
username2 = 'xxxx'        #login name in DD for IP: 10.246.170.162
pswd2 = 'xxxx'        #password for the above login name


#----EDIT THIS-----------^^^^


#Log in  source DD---vvv

command='ssh sysadmin@' + source
print(command)
print('Logging in...')

try:
    child = pexpect.spawn(command)
    i = child.expect (['.*\r\nAccount locked.*', 'Password:','sysadmin@.*# '],timeout=30)
    if i == 0:
        print('Account locked. Can\'t login source DD')
        sys.exit()
    elif i == 1:
        child.sendline(pswd)
        child.expect('sysadmin@.*# ') 
    elif i == 2:
        print('Logged!')
except(pexpect.TIMEOUT):
    print('Can\'t login source DD')
    sys.exit()

#Log in  source DD---^^^



#Log in destination DD---vvv

command='ssh sysadmin@' + destination
print(command)
print('Logging in...')

try:
    child2 = pexpect.spawn(command)
    i = child2.expect (['.*\r\nAccount locked.*', '.*\r\nPassword:.*','sysadmin@.*# '],timeout=20)
    if i == 0:
        print('Account locked. Can\'t login target DD')
        sys.exit()
    elif i == 1:
        child2.sendline(pswd2)
        child2.expect('sysadmin@.*# ')
        print('Logged!')
    elif i == 2:
        print('Logged!')
except(pexpect.TIMEOUT):
    print('Can\'t login target DD')
    sys.exit()

#Log in destination DD---^^^


#Mtree list---vvvvvv
    
child.sendline ('hostname')
child.expect('sysadmin@.*# ')
result = re.findall(r'\w+\.\w+\.\w+', child.before)
print("Mtrees at source DD {0}:".format(result[0]))
child.sendline ('mtree list')
child.expect('sysadmin@.*# ')

mtree_list = child.before.splitlines()
for mtree_string in mtree_list:
    if re.findall(r'^\/data\/col1',mtree_string):
        print(mtree_string)

#Mtree list---^^^^^^

#Creating Mtree-------------------vvvvvv

command='mtree create /data/col1/' + mtree_name
child.sendline (command)
child.expect('sysadmin@.*# ')

#Creating Mtree-------------------^^^^^^

if re.findall(r'.*already exists\.',child.before):
    print('Mtree already exists! Set a new one.')
    child.sendline ('bye')
else:
    #defining src dd hostname
    command='hostname'
    child.sendline (command)
    child.expect('sysadmin@.*# ',timeout=30)
    hostname_src = re.findall(r'\w+\.\w+\.\w+\.\w+',child.before)
    print("Source hostname: {0}".format(hostname_src[0]))

    #defining trgt dd hostname    
    child2.sendline (command)
    child2.expect('sysadmin@.*# ',timeout=30)
    hostname_trgt = re.findall(r'\w+\.\w+\.\w+\.\w+',child2.before)
    print("Destination hostname: {0}".format(hostname_trgt[0]))
    
    if not hostname_trgt or  not hostname_src:
            print('Can\'t  get hostname(s). Exiting...')
            sys.exit()
    
    #setting replication context at src dd
    command='replication add source mtree://' + hostname_src[0] + '/data/col1/' + mtree_name + ' destination ' + 'mtree://' + hostname_trgt[0] + '/data/col1/' + mtree_name
    print(command)
    child.sendline (command)
    child.expect('sysadmin@.*# ',timeout=30)
    print ("To break replication: replication break mtree://{0}/data/col1/{1}".format(hostname_trgt[0],mtree_name))

    
    #setting replication context at trgt dd
    command='replication add source mtree://' + hostname_src[0] + '/data/col1/' + mtree_name + ' destination ' + 'mtree://' + hostname_trgt[0] + '/data/col1/' + mtree_name
    print(command)
    child2.sendline (command)
    child2.expect('sysadmin@.*# ',timeout=30)
    print ("To break replication: replication break mtree://{0}/data/col1/{1}".format(hostname_trgt[0],mtree_name))    
    
    
    #connectivity at src dd and trgt dd, and finally get replication initialized at src dd
    command="replication modify mtree://{0}/data/col1/{1} connection-host {2}".format(hostname_trgt[0], mtree_name, destination) 
    print(command)
    child.sendline (command)
    child.expect('sysadmin@.*# ',timeout=30)
    
    command="replication modify mtree://{0}/data/col1/{1} connection-host {2}".format(hostname_trgt[0], mtree_name, source) 
    print(command)    
    child2.sendline (command)
    child2.expect('sysadmin@.*# ',timeout=30)    

    command="replication initialize mtree://{0}/data/col1/{1}".format(hostname_trgt[0], mtree_name) 
    print(command)    
    child.sendline (command)
    child.expect('sysadmin@.*# ',timeout=30)    
    
    #Final parting words
    print ("\nCompleted!\n")
    message="To watch replication init: replication watch mtree://{0}/data/col1/{1}".format(hostname_trgt[0], mtree_name)
    print(message)
    message="To check out replication status: replication status mtree://{0}/data/col1/{1}".format(hostname_trgt[0], mtree_name)
    print(message)
    message="To export: nfs add /data/col1/{0}/tapelibNAME/DD1_POOL_FS1 192.168.128.0/17 (ro,no_root_squash,all_squash,secure,anonuid=88,anongid=88)".format(mtree_name)
    print(message)
    
    print("\nCleaning after self me - do the following on source and target data domain:\n")
    print("Break replication: replication break mtree://{0}/data/col1/{1}".format(hostname_trgt[0],mtree_name))
    print("Delete MTree: mtree delete /data/col1/{0}".format(mtree_name))
    print('File system clean up: filesys clean start')
    
child.sendline ('bye')
child2.sendline('bye')

File: mtree_replication.pl
#!/usr/bin/perl
use strict;
use warnings;
use Net::SSH::Perl;
use Socket;
#use Data::Dumper;
#use Term::ReadKey;

#==============================================
#19/10/2018
# This script creates mtrees replication between two Data Domain storages
#===============================================


#----EDIT THIS-----------VVVV
#source:
my $mtree_name='HHH';             #Mtree name - a non-existing one
my $source ='x.x.x.x';     #external dd management address for source 
my $username = "xxxx";        #login name in DD for IP: 10.246.168.131
my $pswd = "xxxx";            #password for the above login name

#destination:
my $destination='x.x.x.x'; #external dd management address for destination
my $username2 = "xxxx";       #login name in DD for IP: 10.246.170.162
my $pswd2 = "xxxx";       #password for the above login name

#----EDIT THIS-----------^^^^


my ($stdout, $stderr, $exit);
my $mtree='';
my $source_name='';
my $destination_name='';

my $ssh = Net::SSH::Perl->new($source, passphrase => $pswd, protocol => 2, interactive => 1, use_pty => 1, debug => 0);
$ssh->login($username, $pswd);

my $ssh2 = Net::SSH::Perl->new($destination, passphrase => $pswd2, protocol => 2, interactive => 1, use_pty => 1, debug => 0);
$ssh2->login($username2, $pswd2);

#-----------------------------------
#Listing Mtrees at source
#vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

my $cmd = ('mtree list');
($stdout, $stderr, $exit) = $ssh->cmd($cmd);
print "$stdout\n";

my @values = split(/\n/, $stdout);
foreach my $val (@values) {
  if ($val =~ /^\/data\/col1/) {
   last;
  }
}

#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#Listing Mtrees at source
#-----------------------------------


#-----------------------------------
#Pausing a while to show menu
#vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

print "Mtrees will be created in /data/col1\n";
print "Continue? (y/n):";
my $input = <>;
$input =~ s/[\n\r\f\t]//g;
$input =~ s/^\s+//;

unless ($input eq 'y')
{
exit;
}

#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#Pausing a while to show menu
#-----------------------------------



print "Mtree creating:\n";

#-----------------------------------
#Building Mtree at source
#vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

$cmd = ("mtree create /data/col1/$mtree_name");
($stdout, $stderr, $exit) = $ssh->cmd($cmd);
print "$stdout\n";
print "to delete Mtree: mtree delete /data/col1/XXX\n";
print "to clean orphaned Mtree entries: filesys clean start\n";

#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#Mtree at source
#----------------------------------- 


#-----------------------------------
#Defining the  fully qualified dd hostname for a source and  destination
#vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

$cmd = ("hostname");
($stdout, $stderr, $exit) = $ssh->cmd($cmd);
#print "$stdout\n";

if ($stdout =~ s/^The Hostname is: //) {
   $stdout =~ s/^\s+|\s+$//g;
   $source_name=$stdout;
  }
print "The source name: $source_name\n";


$cmd = ("hostname");
($stdout, $stderr, $exit) = $ssh2->cmd($cmd);

if ($stdout =~ s/^The Hostname is: //) {
   $stdout =~ s/^\s+|\s+$//g;
   $destination_name=$stdout;
  }
print "The destination name: $destination_name\n"; 
  
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#Defining the  fully qualified dd hostname for a source and a destination
#-----------------------------------  


#-----------------------------------
#Creating replication context at source and destination
#vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

$cmd = ("replication add source mtree://$source_name/data/col1/$mtree_name destination mtree://$destination_name/data/col1/$mtree_name");
print "\n$cmd\n";
($stdout, $stderr, $exit) = $ssh->cmd($cmd);
print "To break replication: replication break mtree://$destination_name/data/col1/$mtree_name\n\n";


$cmd = ("replication add source mtree://$source_name/data/col1/$mtree_name destination mtree://$destination_name/data/col1/$mtree_name");
print "$cmd\n";
($stdout, $stderr, $exit) = $ssh2->cmd($cmd);
print "To break replication: replication break mtree://$destination_name/data/col1/$mtree_name\n\n";


#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#Creating replication context at source and destination
#-----------------------------------

#-----------------------------------
#Creating IP connectivity at source and destination, and finally getting it initiated at source
#vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

$cmd = ("replication modify mtree://$destination/data/col1/$mtree_name connection-host $destination");
print "\n\n$cmd\n";
($stdout, $stderr, $exit) = $ssh->cmd($cmd);
print "To break replication: replication break mtree://$destination/data/col1/$mtree_name\n\n";


$cmd = ("replication modify mtree://$destination/data/col1/$mtree_name connection-host $source");
print "$cmd\n";
($stdout, $stderr, $exit) = $ssh2->cmd($cmd);
print "To break replication: replication break mtree://$destination/data/col1/$mtree_name\n\n";


$cmd = ("replication initialize mtree://$destination_name/data/col1/$mtree_name");
print "Now wait...\n\n";
print "$cmd\n";
($stdout, $stderr, $exit) = $ssh->cmd($cmd);

#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#Creating IP connectivity at source and destination, and final initiating it at source
#-----------------------------------

print "\nCompleted\n\n";
print "To watch replication init: replication watch mtree://$destination_name/data/col1/$mtree_name\n";
print "Replication status: replication status mtree://$destination_name/data/col1/$mtree_name\n\n";
print "To delete:\n\n";
print "At destination: replication break  mtree://$destination_name/data/col1/$mtree_name\n";
print "At destination: mtree delete  /data/col1/$mtree_name\n";
print "At destination: filesys clean start\n\n";

print "At source: replication break  mtree://$destination_name/data/col1/$mtree_name\n";
print "At source: mtree delete  /data/col1/$mtree_name\n\n";
print "At source: filesys clean start\n\n";

print "To export: nfs add /data/col1/$mtree_name/tapelibNAME/DD1_POOL_FS1 192.168.128.0/17 (ro,no_root_squash,all_squash,secure,anonuid=88,anongid=88)\n";


